libcloud/compute/drivers/packet.py

```diff
@@ -0,0 +1,254 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Packet Driver
+"""
+
+from libcloud.utils.py3 import httplib
+
+from libcloud.common.base import ConnectionUserAndKey
+from libcloud.common.base import JsonResponse
+from libcloud.compute.types import Provider, NodeState, InvalidCredsError
+from libcloud.compute.base import NodeDriver, Node
+from libcloud.compute.base import NodeImage, NodeSize, NodeLocation
+
+PACKET_ENDPOINT = "api.packet.net"
+
+
+class PacketResponse(JsonResponse):
+    valid_response_codes = [httplib.OK, httplib.ACCEPTED, httplib.CREATED,
+                            httplib.NO_CONTENT]
+
+    def parse_error(self):
+        if self.status == httplib.UNAUTHORIZED:
+            body = self.parse_body()
+            raise InvalidCredsError(body['message'])
+        else:
+            body = self.parse_body()
+            if 'message' in body:
+                error = '%s (code: %s)' % (body['message'], self.status)
+            else:
+                error = body
+            return error
+
+    def success(self):
+        return self.status in self.valid_response_codes
+
+
+class SSHKey(object):
+    def __init__(self, id, label, key):
+        self.id = id
+        self.label = label
+        self.key = key
+
+    def __repr__(self):
+        return (('<SSHKey: id=%s, label=%s, key=%s>') %
+                (self.id, self.label, self.key))
+
+
+class PacketConnection(ConnectionUserAndKey):
+    """
+    Connection class for the Packet driver.
+    """
+
+    host = PACKET_ENDPOINT
+    responseCls = PacketResponse
+
+    def add_default_headers(self, headers):
+        """
+        Add headers that are necessary for every request
+        """
+        headers['Content-Type'] = 'application/json'
+        headers['X-Auth-Token'] = self.user_id
+        headers['X-Consumer-Token'] = \
+            'kcrhMn7hwG8Ceo2hAhGFa2qpxLBvVHxEjS9ue8iqmsNkeeB2iQgMq4dNc1893pYu'
+        return headers
+
+
+class PacketNodeDriver(NodeDriver):
+    """
+    Packet NodeDriver
+    """
+
+    connectionCls = PacketConnection
+    type = Provider.PACKET
+    name = 'Packet'
+    website = 'http://www.packet.net/'
+
+    NODE_STATE_MAP = {'queued': NodeState.PENDING,
+                      'provisioning': NodeState.PENDING,
+                      'rebuilding': NodeState.PENDING,
+                      'powering_on': NodeState.REBOOTING,
+                      'powering_off': NodeState.REBOOTING,
+                      'rebooting': NodeState.REBOOTING,
+                      'inactive': NodeState.STOPPED,
+                      'deleted': NodeState.TERMINATED,
+                      'deprovisioning': NodeState.TERMINATED,
+                      'failed': NodeState.ERROR,
+                      'active': NodeState.RUNNING}
+
+    def list_nodes(self, location):
```

Comment:
so, this is not really a location, it is a project id which is a little unique to how we structure things on the platform. I'll rename the string to not cause confusion here, I hope that works